File: web/backend/tests/test_catalog.py
from rdsa_web.catalog import RobotCatalog


def test_loads_all_robots_and_categories(robots_yaml_path):
    cat = RobotCatalog.from_file(robots_yaml_path)
    assert len(cat.get_all_robots()) == 20
    assert len(cat.get_categories()) == 3


def test_ur3_fields_parsed(robots_yaml_path):
    cat = RobotCatalog.from_file(robots_yaml_path)
    ur3 = cat.get_robot_by_id("ur3")
    assert ur3 is not None
    assert ur3.display_name == "UR3"
    assert ur3.urdf_package == "ur_description"
    assert ur3.urdf_path == "urdf/ur.urdf.xacro"
    assert ur3.xacro_args == "name:=ur3_robot ur_type:=ur3"
    assert ur3.category == "universal_robots"
    assert ur3.specifications.degrees_of_freedom == 6
    assert ur3.specifications.payload_kg == 3.0
    assert ur3.specifications.reach_mm == 500
    assert ur3.specifications.collaborative is True
    assert "collaborative" in ur3.tags
    assert ur3.required_packages == ["ur_description", "ur_msgs"]


def test_mounting_key_maps_to_mounting_options(robots_yaml_path):
    cat = RobotCatalog.from_file(robots_yaml_path)
    ur3 = cat.get_robot_by_id("ur3")
    assert "floor" in ur3.specifications.mounting_options


def test_unknown_robot_id_returns_none(robots_yaml_path):
    cat = RobotCatalog.from_file(robots_yaml_path)
    assert cat.get_robot_by_id("does_not_exist") is None

File: robot_catalog_core/test/test_catalog.cpp
#include <gtest/gtest.h>
#include <cstdlib>
#include "robot_catalog_core/catalog.hpp"
#include "robot_catalog_core/json.hpp"

using robot_catalog::RobotCatalogCore;
using robot_catalog::RobotFilter;
using robot_catalog::RobotConfig;

static std::string yaml_path() {
  const char* p = std::getenv("RDSA_ROBOTS_YAML");
  return p ? p : "";
}

class CatalogTest : public ::testing::Test {
protected:
  RobotCatalogCore cat;
  void SetUp() override { cat.load(yaml_path()); }
};

TEST_F(CatalogTest, LoadsAllRobotsAndCategories) {
  EXPECT_EQ(cat.getAllRobots().size(), 27u);  // 20 arms + 2 EE + 5 bases
  EXPECT_EQ(cat.getCategories().size(), 5u);
}

TEST_F(CatalogTest, Ur3Fields) {
  RobotConfig r;
  ASSERT_TRUE(cat.getRobotById("ur3", r));
  EXPECT_EQ(r.display_name, "UR3");
  EXPECT_EQ(r.urdf_package, "ur_description");
  EXPECT_EQ(r.category, "universal_robots");
  EXPECT_EQ(r.specifications.degrees_of_freedom, 6);
  EXPECT_DOUBLE_EQ(r.specifications.payload_kg, 3.0);
  EXPECT_TRUE(r.specifications.collaborative);
}

TEST_F(CatalogTest, FilterByCategory) {
  RobotFilter f;
  f.category = "universal_robots";
  EXPECT_EQ(cat.filterRobots(f).size(), 9u);
}

TEST_F(CatalogTest, EmptyFilterReturnsAll) {
  EXPECT_EQ(cat.filterRobots(RobotFilter{}).size(), 27u);
}

TEST_F(CatalogTest, SearchCaseInsensitive) {
  EXPECT_EQ(cat.searchRobots("ur3").size(), cat.searchRobots("UR3").size());
  EXPECT_GT(cat.searchRobots("ur3").size(), 0u);
}

TEST_F(CatalogTest, JsonRoundtripShape) {
  auto j = robot_catalog::robots_to_json(cat.getAllRobots());
  EXPECT_NE(j.find("\"id\":\"ur3\""), std::string::npos);
  EXPECT_EQ(j.front(), '[');
}
